Icon translation using PIL

Starbuck5 opened the pull request:
Fixes #6510

This puts a little extra code in the Mac path and the Windows path to try and convert icons into the correct platform format using PIL/Pillow.

The code for the paths is similar, so it could potentially be centralized in one location, but that seemed more disruptive, especially since I don't know the PyInstaller internals.

One thing I noticed is that the old Mac path didn't fail on incorrect icons, it would just transfer them into the bundle and MacOS wouldn't recognize the images properly. It now will fail explicitly like Windows does.

I tried to do the complete code and docs, but I haven't yet done the tests. I was unable to run the tests on Mac, because of #6430. I did get stuff building so I could see the icon, by making a couple codesigning-adjacent functions in osx.py just return instantly, but I don't think I got a functional executable out of it.

On Windows, I get: 
`FAILED tests/functional/test_basic.py::test_invalid_icon - KeyError: 'cachedir'`
As well as
`FAILED tests/unit/test_modulegraph/test_zipio.py::TestModuleGraph::test_islink - OSError: [WinError 1314] A required ...FAILED tests/unit/test_modulegraph/test_zipio.py::TestModuleGraph::test_readlink - OSError: [WinError 1314] A require...`
Which I assume is flakiness / windows defender, since I don't see how that could come from this PR.

I'm unsure how to properly test this, given PIL / no PIL.

Starbuck5 committed:
PIL icon translation [Windows]

Starbuck5 committed:
PIL icon translation [Doc]

Starbuck5 committed:
PIL icon translation [Mac]

bwoodsend commented on PyInstaller/building/osx.py:
```diff
@@ -17,15 +17,14 @@
 from PyInstaller.building.datastruct import TOC, Target, logger
 from PyInstaller.building.utils import (_check_path_overlap, _rmtree, add_suffix_to_extension, checkCache)
 from PyInstaller.compat import is_darwin
+from PyInstaller.config import CONF
```
Ah, see this is one of the many reasons I hate the `CONF` object. The import genuinely does have to be inside `Bundle.__init__()` so that it is updated whenever we do nonsense like [this](https://github.com/pyinstaller/pyinstaller/blob/develop/tests/unit/test_hookutils.py#L113) to work around it. I believe this is the cause of the `KeyError`s you're getting.

Starbuck5 replied:
The only `KeyError` I see is in the Windows test suite, because I didn't remove the old test, which called into the building code directly with some assumptions. Unless I'm misreading the Mac test logs. I'm not sure why I'm getting 7 fails on Mac, it doesn't seem to give a reason.

bwoodsend commented:
Nice work. I think you want `CONF['workpath']` (the `build` directory) rather than `CONF["cachedir"]` (the global cache).

Starbuck5 committed:
lint, cachedir->workpath, CONF import, remove old test

Starbuck5 commented:
>  Nice work. I think you want CONF['workpath'] (the build directory) rather than CONF["cachedir"] (the global cache).

I was wondering about that-- I just wanted to put the generated image somewhere out of the way so I didn't have to worry about deleting it. `build/` makes sense.

I'm still very uncertain about how I should test this. How do the splash screen tests react to the presence / non presence of PIL? Should the test hit Mac and Windows the same way, since they should react the same, and do a full build? Or should it be like the old test, where it injects into the specific method and tests the response?

Also I guess I need to put an entry in for towncrier, if that's still the process.

rokm commented:
The two conversion codepaths look quite similar in structure - do you think it would be possible to merge them into a single helper function? E.g., one that takes the input icon name and returns either the original input name (if icon was in supported format) or convert the icon and return the new filename. And the differences could be parametrized as well, I suppose (e.g., list of OS-supported formats, the OS name in error message, etc.).

Then we could test just the behavior of this helper function in the context of a unit test instead of full build test (where we have limited options of inspecting the result, unless we opt for checking the (generated) icon file in the `build` directory)...

Starbuck5 commented:
@rokm That's a good idea for centralizing code and for testing.

Do you have any location in mind for where the utility function should go? I see utils/misc.py, but maybe there's a place that fits better.

bwoodsend commented:
I'd prefer that it goes in the file it gets used in rather than buried in one of our many `utils` submodules.

rokm commented:
> I'd prefer that it goes in the file it gets used in rather than buried in one of our many `utils` submodules.

But if the helper needs to be called from two different files, then I guess we can live with a new `PyInstaller.utils.icon` submodule (and a `tests/units/test_icon.py` for unit test).

Starbuck5 committed:
Move to centralized function windows

Starbuck5 committed:
Try for mac

I somehow can't install pyinstaller on this anymore, it's really quite strange.

bwoodsend commented on PyInstaller/utils/win32/icon.py:
```diff
@@ -202,7 +203,13 @@ def splitter(s):
+
+    # validates the icon, which means it will always return a path to an existing "exe"
+    # or "ico" file or raise an Exception trying
+    srcpath = validate_icon(srcpath, ("exe", "ico"), "ico", config.CONF["workpath"])
```
Maybe rename `validate_icon()` to `normalise_icon_type()` or `as_native_icon()` - then we shouldn't need the comment above?

Starbuck5 replied:
Naming is hard 😄 

`as_native_icon` was really growing on me but then I realized its not a verb, so it seems wrong for a function name.

I haven't changed it yet.

Starbuck5 replied:
I changed it to `normalize_icon_type`.

bwoodsend commented on PyInstaller/utils/icon.py:
```diff
@@ -0,0 +1,56 @@
+from typing import Tuple
```
Can we make this `PyInstaller/building/icon.py` since in a more sensibly laid out repository, executable assembly would be the only place this should get used?

bwoodsend commented on PyInstaller/utils/icon.py:
```diff
@@ -0,0 +1,56 @@
+            raise ValueError(
+                f"Received icon image '{icon_path}' which exists but is not in the correct format. On this platform, only {allowed_types} "
+                f"images may be used as icons. If PIL is installed, automatic conversion "
+                f"will be attempted. Please install PIL or convert your '{extension}' file to one of {allowed_types} "
```
Call it `pillow` here since that's what the user should `pip install`.

bwoodsend commented on PyInstaller/utils/icon.py:
```diff
@@ -0,0 +1,56 @@
+from typing import Tuple
```
We also need the copyright header. GPL rules I'm afraid :roll_eyes:

Starbuck5 replied:
Got it.

bwoodsend commented on PyInstaller/utils/icon.py:
```diff
@@ -0,0 +1,56 @@
+
+        if PILImage:
+            try:
+                generated_icon = os.path.join(workpath, f"generated.{convert_type}")
```
I'm not sure but in a really nasty case where the user is using `MERGE()` with `EXE()`s with different icons or is running multiple builds concurrently then I think having those two icons both written to `$workpath/generated.$suffix` could lead to them getting mixed up. Maybe squirt a hash in there to keep different source images unique?
```python
                _generated_name = f"generated-{hashlib.sha256(icon_path.encode()).hexdigest()}{convert_type}"
                generated_icon = os.path.join(workpath, _generated_name)
```

Starbuck5 replied:
While I was thinking about this a couple hours ago I put in the process ID of the running python process to deal with concurrency. But coming back I see that could still be a problem with multiple icons in the same process, so I'll revise that.

Starbuck5 replied:
Ok, I now simply do "generated + hash(file path)"

bwoodsend replied:
The problem with the builtin `hash()` is that it's randomly seeded on Python startup.
```python
> python
>>> hash("hello")
3685732701264951447
>>> hash("hello")
3685732701264951447

> python
>>> hash("hello")
-8441502141078742039
>>> hash("hello")
-8441502141078742039
```
So if someone rebuilds their application repeatedly then they'll end up with hundreds of copies of this converted icon in their `build` folder.

Starbuck5 replied:
Ah, that's a good catch. I put in your hashlib strategy.

bwoodsend commented on PyInstaller/utils/icon.py:
```diff
@@ -0,0 +1,56 @@
+
+def validate_icon(icon_path: str, allowed_types: Tuple[str], convert_type: str, workpath: str) -> str:
+    """
+    Outputs or a valid icon path or raises an Exception trying
```
Outputs or a valid icon?

Starbuck5 replied:
Good catch.

bwoodsend reviewed (commented):
I would like to see some kind of unit test. Just something simple that passes some images (the ones in `PyInstaller/bootloader/images/` would do) through `validate_icon()` and covers the different cases:

* Native image - file path is passed through unchanged
* PNG - output is a different file with the correct suffix
* Some random non-image file: Raises an image conversion error
* A PNG after calling `monkeypatch.setitem(sys.modules, "PIL", None)`: Raise the _install pillow_ error

No need to actually verify that the images it produces are valid since that's pillow's code we'd be testing - not ours.

Starbuck5 committed:
review feedback

Starbuck5 committed:
Test!

Starbuck5 committed:
Lint

Starbuck5 committed:
Rename to normalize_icon_type

Also changes how the generated file is named to hopefully better avoid conflict

bwoodsend commented on tests/unit/test_normalize_icon_type.py:
```diff
@@ -0,0 +1,65 @@
+
+def test_normalize_icon_type(monkeypatch):
+    from PyInstaller.building.icon import normalize_icon_type
+
+    # Currently assumes PIL/Pillow is installed on the testing machine
```
```suggestion
    pytest.importorskip("PIL")
```
Not any more!

Starbuck5 replied:
I put it down a bit so it could still fail earlier on the tests that don't need PIL.

bwoodsend commented on tests/unit/test_normalize_icon_type.py:
```diff
@@ -0,0 +1,65 @@
+
+    # Currently assumes PIL/Pillow is installed on the testing machine
+
+    workpath = data_dir = "PyInstaller/bootloader/images"
```
`data_dir` needs to be made independent of working directory and PyInstaller installation location. That's why the Alpine test is failing. The `workpath` should be a temporary directory (see [tmp_path fixture](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html#tmp-path)) so that we a) have test isolation and b) aren't in danger of accidentally uploading those generated images to PyPI the next time we do a release. Also, this would all be a lot neater using `pathlib.Path()` instead of `os.path`.


```suggestion
    from pathlib import Path
    import PyInstaller
    data_dir = Path(PyInstaller.__file__).with_name("bootloader") / "images"
```

Starbuck5 replied:
I think I've got it now.

bwoodsend replied:
You're rather missing the point of `pathlib`. The idea is that you leave `data_dir` as a `Path()` then later on you can use `icon = data_dir / 'icon-console.ico'` instead of `icon = os.path.join(data_dir, 'icon-console.ico')`. But that doesn't really matter.

Starbuck5 replied:
I am new to pathlib, but I wanted to make sure the paths passed to the normalize function are strings, as that's what they will be in use.

Maybe pathlib paths are the same as strings for most things, but they aren't totally equivalent. In my time as a pygame contributor we've we had to put some special case code in to deal with Paths.

bwoodsend reviewed (changes requested):
Nearly there...

Starbuck5 committed:
review feedback 2

Starbuck5 commented:
When this is ready I should probably do some git squashing to fix up the git history before it is merged.

bwoodsend commented:
> When this is ready I should probably do some git squashing to fix up the git history before it is merged.

I was intending to squash the lot into one anyway. I prefer docs and tests to go into the same commit as their feature and the rest of the commits are just amendments which we don't need a permanent log of.

rokm commented on PyInstaller/building/icon.py:
```diff
@@ -0,0 +1,70 @@
+#-----------------------------------------------------------------------------
+# Copyright (c) 2013-2022, PyInstaller Development Team.
```
As this is new code, the year should probably be 2022 only.

rokm commented on PyInstaller/building/icon.py:
```diff
@@ -0,0 +1,70 @@
+    _, extension = os.path.splitext(icon_path)
+    extension = extension[1:]  # get rid of the "." in ".whatever"
+
+    if extension not in allowed_types:
```
I think this would better be inverted; if `extension` is in `allowed_types`, return the original path. Else, do the conversion. This way, we can get rid of one level of nesting.

Starbuck5 replied:
I reworked the logic a bit based on your suggestions, it is nice to get rid of one level of nesting.

rokm commented on PyInstaller/building/icon.py:
```diff
@@ -0,0 +1,70 @@
+            from PIL import Image as PILImage
+            import PIL
+        except ImportError:
+            PILImage = None
```
Why not raise the `ValueError` here immediately?

rokm commented on PyInstaller/building/osx.py:
```diff
@@ -123,11 +124,14 @@ def assemble(self):
+        # Makes sure the icon exists and attempts to convert to the proper format if applicable
+        self.icon = normalize_icon_type(self.icon, ("icns",), "icns", CONF["workpath"])
+
+        #Ensure icon path is absolute
```
Space after #.

Starbuck5 committed:
review feedback 3

Starbuck5 committed:
missed one

Starbuck5 commented:
> I was intending to squash the lot into one anyway. I prefer docs and tests to go into the same commit as their feature and the rest of the commits are just amendments which we don't need a permanent log of.

Good old squash and merge

rokm commented on PyInstaller/building/icon.py:
```diff
@@ -0,0 +1,70 @@
+
+def normalize_icon_type(icon_path: str, allowed_types: Tuple[str], convert_type: str, workpath: str) -> str:
+    """
+    Outputs a valid icon path or raises an Exception trying
```
Returns a valid icon path or raises an Exception on error.

rokm commented on PyInstaller/building/icon.py:
```diff
@@ -0,0 +1,70 @@
+def normalize_icon_type(icon_path: str, allowed_types: Tuple[str], convert_type: str, workpath: str) -> str:
+    """
+    Outputs a valid icon path or raises an Exception trying
+    Checks to be sure the icon exists, and attempts to use Pillow to convert
```
Ensures that the icon exists, and, if necessary, attempts to convert it to correct OS-specific format using Pillow.

rokm commented on PyInstaller/building/icon.py:
```diff
@@ -0,0 +1,70 @@
+    it to the right format if it's in an unsupported format
+
+    Takes:
+    icon_path - the icon entered by the user
```
entered -> given

rokm commented on PyInstaller/building/icon.py:
```diff
@@ -0,0 +1,70 @@
+        from PIL import Image as PILImage
+        import PIL
+
+        try:
```
We can avoid double try block here. Have the first try block handle imports (and raise `ValueError` on `Import error`), and have the second try block handle conversion itself.

rokm commented on news/6697.feature.rst:
```diff
@@ -0,0 +1,2 @@
+Translate icon images into the correct platform format automatically
+using PIL/Pillow, if installed.
```
Automatically convert icons into platform-supported format using PIL/Pillow, if available.

rokm commented on tests/unit/test_normalize_icon_type.py:
```diff
@@ -0,0 +1,67 @@
+import PyInstaller
+
+
+def test_normalize_icon_type(monkeypatch, tmp_path):
```
I think the single test should be split into multiple separate tests, so that the non-Pillow ones get marked as passed instead of skipped when Pillow is unavailable:
- file-not-found error test
- native image
- converted image
- random image
- install-the-pillow error

Starbuck5 replied:
Oh, I misread this. I separated the tests out into two groups, one of which needs PIL and one of which doesn't.

(I think that's okay, it accomplishes the objective)

rokm replied:
I'd prefer to see them separate, but it's not a deal-breaker at this point. The current split does accomplish the objective, and we can split them up later if necessary.

rokm reviewed (changes requested):
Nice work! I have couple more nitpicks, then I think this should be ready.

Starbuck5 committed:
review feedback 4

Starbuck5 committed:
de-nest the internal try block

bwoodsend commented:
No need to panic about that Windows CI failure - it's happening elsewhere too (although we haven't figured out why).

bwoodsend commented:
You made it! Nice one @Starbuck5
